credentials/xds/xds.go

```diff
@@ -0,0 +1,339 @@
+// security configuration is pushed by a management server using xDS APIs.
+//
+// All APIs in this package are EXPERIMENTAL.
+package xds
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	credinternal "google.golang.org/grpc/internal/credentials"
+)
+
+// ClientOptions contains parameters to configure a new client-side xDS
+// credentials implementation.
+type ClientOptions struct {
+	// FallbackCreds specifies the fallback credentials to be used when either
+	// the `xds` scheme is not used in the user's dial target or when the xDS
+	// server does not return any security configuration. Attempts to create
+	// client credentials without a fallback credentials will fail.
+	FallbackCreds credentials.TransportCredentials
+}
+
+// NewClientCredentials returns a new client-side transport credentials
+// implementation which uses xDS APIs to fetch its security configuration.
+func NewClientCredentials(opts ClientOptions) (credentials.TransportCredentials, error) {
+	if opts.FallbackCreds == nil {
+		return nil, errors.New("missing fallback credentials")
+	}
+	return &credsImpl{
+		isClient: true,
+		fallback: opts.FallbackCreds,
+	}, nil
+}
+
+// credsImpl is an implementation of the credentials.TransportCredentials
+// interface which uses xDS APIs to fetch its security configuration.
+type credsImpl struct {
+	isClient bool
+	fallback credentials.TransportCredentials
+}
+
+// handshakeCtxKey is the context key used to store HandshakeInfo values.
+type handshakeCtxKey struct{}
+
+// HandshakeInfo wraps all the security configuration required by client and
+// server handshake methods in credsImpl. The xDS implementation will be
+// responsible for populating these fields.
+//
+// Safe for concurrent access.
+type HandshakeInfo struct {
+	mu               sync.Mutex
+	rootProvider     certprovider.Provider
+	identityProvider certprovider.Provider
+	acceptedSANs     []string // Only on the client side.
+}
+
+// SetRootCertProvider updates the root certificate provider.
+func (chi *HandshakeInfo) SetRootCertProvider(root certprovider.Provider) {
+	chi.mu.Lock()
+	chi.rootProvider = root
+	chi.mu.Unlock()
+}
+
+// SetIdentityCertProvider updates the identity certificate provider.
+func (chi *HandshakeInfo) SetIdentityCertProvider(identity certprovider.Provider) {
+	chi.mu.Lock()
+	chi.identityProvider = identity
+	chi.mu.Unlock()
+}
+
+// SetAcceptedSANs updates the list of accepted SANs.
+func (chi *HandshakeInfo) SetAcceptedSANs(sans []string) {
+	chi.mu.Lock()
+	chi.acceptedSANs = sans
+	chi.mu.Unlock()
+}
+
+func (chi *HandshakeInfo) validate(isClient bool) error {
+	chi.mu.Lock()
+	defer chi.mu.Unlock()
+
+	// On the client side, rootProvider is mandatory. IdentityProvider is
+	// optional based on whether the client is doing TLS or mTLS.
```

Comment:
From the comment, I guess we might also need a variable for users to choose between TLS and mTLS?

Reply:
The control plane sends two `plugin_instance_names` which are resolved to the plugin configuration by looking up a bootstrap file. One of the names resolves to the configuration to be used to fetch the identity cert, while the other name resolves to the configuration to be used to fetch the root certificate. If the former is absent, we do TLS and if the former is present, we do mTLS.